parsl/executors/high_throughput/zmq_pipes.py

```diff
@@ -24,9 +27,15 @@ def __init__(self, ip_address, port_range):
         self.port = self.zmq_socket.bind_to_random_port("tcp://{}".format(ip_address),
                                                         min_port=port_range[0],
                                                         max_port=port_range[1])
+        self.poller = zmq.Poller()
+        self.poller.register(self.zmq_socket, zmq.POLLOUT)
 
     def put(self, message):
```

Comment:
This is (now more explicitly) blocking (for the zmq_socket to be ready), which causes HighThroughputExecutor.submit to be similarly blocking.

The documentation for concurrent.futures.Executor suggests (but is not explicit) that submit should be non-blocking, and return a future "quickly"; and other code possibly relies on that behaviour.

What are the conditions in which this code will block?

Should there be a logged warning if we are blocking here? For example, if the first poll does not return the socket as available?

Reply:
Correct. This call is now a blocking call to ensure that the receiving end has space for an incoming message. If we do not do this, the receiver can overflow and the socket can go bad. There's a debug message that now explains why some tasks take a few more ms to go out.

But moving to this model means that we can still sustain 2000Tasks/s pump rate without having the receiver fall over.